napari/_vispy/layers/labels.py

```diff
@@ -478,24 +511,27 @@ def _on_colormap_change(self, event=None):
         mode = self.layer.color_mode
 
         if mode == 'auto':
-            dtype = minimum_dtype_for_labels(self.layer.num_colors + 1)
-            if issubclass(dtype.type, np.integer):
-                scale = np.iinfo(dtype).max
-            else:  # float32 texture
-                scale = 1.0
+            data_dtype = self.layer._slice.image.view.dtype
+            raw_dtype = self.layer._slice.image.raw.dtype
+            if data_dtype != raw_dtype:
+                colormap = LabelColormap(**colormap.dict())
+                colormap.background_value = _cast_labels_to_minimum_dtype_auto(
+                    np.array([colormap.background_value]).astype(raw_dtype),
+                    colormap,
+                )[0]
+            if data_dtype == np.uint8:
+                colors = colormap._uint8_colors.reshape(256, -1, 4)
+            else:
+                colors = colormap._uint16_colors.reshape(256, -1, 4)
```

Comment:
Maybe call this `color_texture`? This makes it clearer why we are reshaping and why the colors are not themselves just stored in the texture shape from the beginning.